### PR TITLE
New Foundation

### PR DESCRIPTION
**Pull request by leon927**
Just a quick update with a pull request. Still WIP and a lot of things to finalize but it is going really well! Initial tests are great!

Solved a lot of unknown bugs so far. Not that noticeable but they were there. Also optimized several parts so they execute/calculate faster. The resolution/response seems to be much better just by improving calculations and parameters. Overall much safer, especially for future development. The entire controller code is also much simpler and improved in so many ways.

Still, this is not even an alpha version. Have more things to do so will develop for another few days.

### CODE REVIEW

File: src/controller/adc.h [hunk 1]
```diff
@@ -17,10 +17,21 @@
 #define UI8_ADC_THROTTLE 				      (*(uint8_t*)(0x53EE)) // AIN7
 #define UI8_ADC_TORQUE_SENSOR         (*(uint8_t*)(0x53E8)) // AIN4
 
+#define UI16_ADC_10_BIT_BATTERY_VOLTAGE 	(((*(uint8_t*)(0x53EC)) << 2) | (*(uint8_t*)(0x53ED)))
```
Comment (Frans-Willem) [inline src/controller/adc.h:21]: I'd suggest adding a cast (uint16_t)(...) over the final value, such that if you try to assign it to an uint8_t, the compiler will warn you.

File: src/controller/main.h [hunk 2]
```diff
@@ -12,15 +12,37 @@
...
-#define MIDDLE_PWM_DUTY_CYCLE_MAX                 (PWM_DUTY_CYCLE_MAX/2)
 
 
-// motor rotor offset 
-#define MOTOR_ROTOR_OFFSET_ANGLE                  10
+// motor 
+#define PWM_CYCLES_COUNTER_MAX                                    3125    // 5 erps minimum speed; 1/5 = 200ms; 200ms/64us = 3125
+#define PWM_CYCLES_SECOND                                         15625L  // 1 / 64us(PWM period)
+#define PWM_DUTY_CYCLE_MAX                                        254
+#define PWM_DUTY_CYCLE_MIN                                        0
```
Comment (Frans-Willem) [inline src/controller/main.h]: Is this safe ?

File: src/controller/main.h [hunk 3]
```diff
@@ -12,15 +12,37 @@
...
+  A higher value will make the motor acceleration slower.
+  
+  For a 24 V battery, 25 for ramp up seems ok. For a 
+  higher voltage battery, this values should be higher.
+---------------------------------------------------------*/
+
+
+
+#define MOTOR_ROTOR_OFFSET_ANGLE                                  10
+#define MOTOR_ROTOR_ANGLE_90                                      73      // (63  + MOTOR_ROTOR_OFFSET_ANGLE)
```
Comment (Frans-Willem) [inline src/controller/main.h]: Just to clarify again, does SDCC really not optimize things like (63 + MOTOR_ROTOR_OFFSET_ANGLE) ?

File: src/controller/main.h [hunk 4]
```diff
@@ -12,15 +12,37 @@
...
-// motor rotor offset 
-#define MOTOR_ROTOR_OFFSET_ANGLE                  10
+// motor 
+#define PWM_CYCLES_COUNTER_MAX                                    3125    // 5 erps minimum speed; 1/5 = 200ms; 200ms/64us = 3125
+#define PWM_CYCLES_SECOND                                         15625L  // 1 / 64us(PWM period)
+#define PWM_DUTY_CYCLE_MAX                                        254
+#define PWM_DUTY_CYCLE_MIN                                        0
+#define MIDDLE_PWM_DUTY_CYCLE_MAX                                 127     // PWM_DUTY_CYCLE_MAX / 2
+
+#define PWM_DUTY_CYCLE_RAMP_UP_INVERSE_STEP                       60      // 60 -> 60 * 64 us for every duty cycle increment
```
Comment (Frans-Willem) [inline src/controller/main.h]: Why did you change this from 20 to a higher value ?

File: src/controller/motor.c [hunk 5]
```diff
@@ -622,15 +606,17 @@ void TIM1_CAP_COM_IRQHandler(void) __interrupt(TIM1_CAP_COM_IRQHANDLER)
...
   // do not control current at every PWM cycle, that will measure and control too fast. Use counter to limit 
   if (++ui8_current_controller_counter > 12)
   {
     // reset counter
     ui8_current_controller_counter = 0;
     
-    // if battery max current or phase current is too much, reduce duty cycle
-    if ((ui8_g_adc_battery_current > ui8_controller_adc_battery_max_current) || (ui8_adc_motor_phase_current > ui8_adc_target_motor_phase_max_current))
+    // if battery max current or motor phase current is too high, reduce duty cycle
+    if ((ui8_g_adc_battery_current > ui8_controller_adc_battery_current_target) || (ui8_adc_motor_phase_current > ui8_adc_target_motor_phase_max_current))
```
Comment (Frans-Willem) [inline src/controller/motor.c]: Are you absolutely certain that you want to remove the check to see if the battery current is more than the maximum allowed?
Did you maybe intend to add a check, instead of replacing one ?

File: src/controller/motor.c [hunk 6]
```diff
@@ -657,12 +643,12 @@ void TIM1_CAP_COM_IRQHandler(void) __interrupt(TIM1_CAP_COM_IRQHANDLER)
   }
   else // nothing to limit, so adjust duty_cycle to duty_cycle_target, including ramping
   {
-    // if brake is active, keep duty cycle target at 0
-    if (ui8_motor_controller_state & MOTOR_CONTROLLER_STATE_BRAKE) { ui8_controller_duty_cycle_target = 0; }
```
Comment (Frans-Willem) [inline src/controller/motor.c:661]: Are you sure removing this check is safe ?

File: src/controller/motor.c [hunk 7]
```diff
@@ -657,12 +643,12 @@ void TIM1_CAP_COM_IRQHandler(void) __interrupt(TIM1_CAP_COM_IRQHANDLER)
...
-    
     if (ui8_controller_duty_cycle_target > ui8_g_duty_cycle)
     {
-      if (ui16_counter_duty_cycle_ramp_up++ >= ui16_duty_cycle_ramp_up_inverse_step)
+      // limit duty cycle ramp up
+      if (ui16_duty_cycle_ramp_up_inverse_step < 20 ) { ui16_duty_cycle_ramp_up_inverse_step = 20; }
```
Comment (Frans-Willem) [inline src/controller/motor.c]: Huh?
You've deliberately increased the define that sets this variable to 60, and now you're capping it at 20?

File: src/controller/brake.c [hunk 8]
```diff
@@ -18,25 +18,25 @@
...
+  // initialize the interrupt sensitivity
   EXTI_SetExtIntSensitivity(EXTI_PORT_GPIOC, EXTI_SENSITIVITY_RISE_FALL);
 }
 
 
 // brake signal interrupt
 void EXTI_PORTC_IRQHandler(void) __interrupt(EXTI_PORTC_IRQHANDLER)
 {
-  if (brake_is_set())
+/*   if (brake_is_set())
```
Comment (Frans-Willem) [inline src/controller/brake.c:32]: Are we sure removing all this brake related code is safe?
Are you running brakes on your test system ?

File: src/controller/ebike_app.c [hunk 9]
```diff
@@ -118,77 +106,89 @@ static void check_brakes(void);
...
   
-  // throttle
-  apply_throttle();
+  // torque assist
+  apply_torque_assist();
   
-  // cruise
-  apply_cruise();
+  // eMTB
+  apply_emtb();
```
Comment (Frans-Willem) [inline src/controller/ebike_app.c]: All these functions seem to check the riding mode first.
Wouldn't it be more legible to just have a switch case here checking the riding mode, and then calling the relevant apply_...() function ?

**Review by Frans-Willem — COMMENTED** [review]
Haven't been able to check everything. In the future if you'd like more useful feedback/reviews, it'd be best to split these kinds of things out into separate pull requests.
E.g. the renames of certain variables (wheel_max_speed <-> wheel_speed_max) could've been a single pull request that would've been really easy to review, and would remove a lot of the "noise" in this one.

File: src/controller/adc.h (hunk 1 above)
Comment (leon927) [reply src/controller/adc.h:21]: Great suggestion! However not sure if this define will be used. Will keep it in mind nonetheless, thanks!

File: src/controller/ebike_app.c (hunk 9 above)
Comment (leon927) [reply src/controller/ebike_app.c]: Doing a lot of things at the same time. Compare the code to 0.19.0 and you will see what I mean. It will be improved when I have checked my priority items.

File: src/controller/brake.c (hunk 8 above)
Comment (leon927) [reply src/controller/brake.c:32]: WIP.

File: src/controller/motor.c (hunk 7 above)
Comment (leon927) [reply src/controller/motor.c]: If it is less than 20 it will be set to 20. It is correct. Note that it is the inverse step.

File: src/controller/motor.c (hunk 6 above)
Comment (leon927) [reply src/controller/motor.c:661]: WIP.

File: src/controller/motor.c (hunk 5 above)
Comment (leon927) [reply src/controller/motor.c]: Did not remove anything. Also not the max, it is the target.

File: src/controller/main.h (hunk 4 above)
Comment (leon927) [reply src/controller/main.h]: WIP.

File: src/controller/main.h (hunk 3 above)
Comment (leon927) [reply src/controller/main.h]: Even if it does optimize things like that, it does not hurt to change it. I trust Casainho when he says that differences such as these make a difference when using SDCC.

File: src/controller/main.h (hunk 2 above)
Comment (leon927) [reply src/controller/main.h]: Yes.

**Comment by leon927** [comment]
> 
> 
> Haven't been able to check everything. In the future if you'd like more useful feedback/reviews, it'd be best to split these kinds of things out into separate pull requests.
> E.g. the renames of certain variables (wheel_max_speed <-> wheel_speed_max) could've been a single pull request that would've been really easy to review, and would remove a lot of the "noise" in this one.

When doing rapid development and trying to be effective with my time it is not that optimal to create separate pull requests for each and every step. Some things I am testing purely for development and these things will be removed. Making it very unnecessary to add in a pull request and discuss. Also, the more I discuss why I do things for the moment the more time I lose for development. 

I am forever grateful for feedback but would appreciate it more when I am getting ready to release it to users. This is still in heavy development for at least a couple of days.

Some things and changes are made so the development on the new displays will be much easier for new developers. I would LOVE it if more developers would start developing for this great community so that we get support for more displays. But for this to happen I want the controller code to be the very best it can be. I for one would have nothing to do except develop display code if I am able to make this controller code "perfect".

Casainho has done a fantastic job with the foundation and my only goal is to improve where possible and in some way finalize the controller code. It will then be much simpler to develop the code for the displays.

File: src/controller/motor.c (hunk 7 above)
Comment (Frans-Willem) [reply src/controller/motor.c]: Ah sorry, my bad, misread the comparison :)

**Comment by Frans-Willem** [comment]
I understand it's usually a hassle to split out commits and add proper commit messages, but if we'd like to get more developers involved, I think this is important.

From my perspective: I've got a few things I'd like to change, and I'd like to dive in and add them to the code. However the main branch has already diverged quite a bit from 19.0, and all of the commits are huge commits with changes all over the place, with mostly empty commit messages like "WIP".

How am I expected to figure out what's changed functionally, if it's even safe to run the main branch on my motor, or what's currently "in progress" and what's "stable" ?
If I actually do start coding, how do you expect me to solve merge conflicts if I can't even find out why things were changed in the first place ?

So yeah, clean commits and commit messages definitely are a hassle, but they're invaluable when working together with other people. If it's only your own personal project, by all means do whatever works for you, but if we'd like to get other people involved, we need to make sure they have the information they need to actually get involved.

File: src/controller/main.h (hunk 3 above)
Comment (Frans-Willem) [reply src/controller/main.h]: If it actually makes a difference in the generated code by SDCC, by all means we should change these constants.
The downside of defining constants that depend on other constants like this, is that they get out of sync. e.g. if someone changes MOTOR_ROTOR_OFFSET_ANGLE, how will you ensure they don't forget to change all the other related defines ?
In this case they're right next to each other, but what if there are other defines depending on this in other header files ?
That's why, if SDCC does actually do proper constant propagation, it's probably better to leave the calculations in.

(I'll check later to see when and where SDCC does or doesn't do constant propagation to actually get this discussion settled ;) )

File: src/controller/motor.c (hunk 5 above)
Comment (Frans-Willem) [reply src/controller/motor.c]: In the old code,  there's a check reading  "battery_current > battery_max_current", presumably to protect against an over-current situation.
In the new code, you're only comparing to current_target, no longer to the maximum.

File: src/controller/main.h (hunk 3 above)
Comment (leon927) [reply src/controller/main.h]: As they are close to each other it is okay in this case. But this is NOT something one should change if not knowing what they are changing. They should know to change the other values as well. 

Please do check if SDCC optimizes things like this and settle the discussion :)

File: src/controller/motor.c (hunk 5 above)
Comment (leon927) [reply src/controller/motor.c]: Same thing and same checks. Please look at the code more closely if concerned.

File: src/controller/motor.c (hunk 7 above)
Comment (leon927) [reply src/controller/motor.c]: No problem at all!

**Comment by leon927** [comment]
The commits are huge as there are a lot of changes being made and some are not incorporated whilst others are. Many things are dependent so there is no chance of not changing a lot of things.

I also did not know that there are other people wanting to develop for the controller. I have been active for a couple of months now and there have been no developers, except Casainho, showing interest nor intent of doing some official changes and actual development.

I do agree that regular development should be with clear documentation. But in this case it is basically a complete rewrite on many things and should be handled as such.

If you are eager to develop would you mind being patient for the time being? Because a lot of things had to be changed in order for us to have multiple riding modes and functions. At the same time have an easy path for adding support for more displays.

If you look at the past you will see smaller pull requests but that is normal development. This is basically me trying to do the best I can to simplify everything so developers such as yourself can make changes more easily. The code has gotten a lot simpler so anyone can make changes without it causing strange bugs throughout.

**Comment by Frans-Willem** [comment]
I appreciate you trying to simplify things for others, but in the process you're making it very hard for me to pitch in. Basically the current master branch as well as this pull request is a combination of refactoring for simplicity, stuff that's still in progress, as well as new untested and undocumented features (some done, some not), all mixed together in an incoherent collection of commits :-/

I think for the time being, I'll just branch off of 19.0 and implement the fixes that I need. If you'd like other people to contribute on the main branch, I'd highly recommend to:
- Keep commits as either refactoring or new features (not both in the same commit)
- Keep work-in-progress in separate branches clearly marked as unstable.

Furthermore, if simplicity is what you're after, it's paramount that you discuss code with others before committing to it. What may appear simple when you're knee-deep in the code and working on that specific feature, may not appear simple at all to someone else reading the code.

**Comment by leon927** [comment]
Really sorry you feel that the rewrite I am doing to simplify and improve the entire code should be carried out in a different way. This is not to be compared to regular development. 

Anyway, your plan for the time being is solid. Soon there will also be a beta and you are free to give feedback and suggestions then. Hopefully you will like the changes and the issues you reported will be solved or at least significantly reduced.

My intention is not to make it harder for you or in any way take away from your experience with this community, so I truly do feel sorry you feel that way.

Just want to add another reason I am rapidly developing and that is because everything should be somewhat finished with a good layout for the introduction of the new displays. Casainho has said he wishes to make a working prototype of the SW102 within a couple of weeks and I am trying to beat that deadline. I do not have that display at the moment but I know when it is time to develop to that display I would want to have the changes in the controller I have implemented so far.

File: src/controller/main.h (hunk 3 above)
Comment (Frans-Willem) [reply src/controller/main.h]: Just checked on 0.19.0, compiled once with the old style (MOTOR_ROTOR_OFFSET + something), and once with constants pre-calculated. Both result in the exact same hexfile.

Full output:
```
[frans-willem@Hydra controller]$ sdcc --version
SDCC : mcs51/z80/z180/r2k/r3ka/gbz80/tlcs90/ez80_z80/ds390/pic16/pic14/TININative/ds400/hc08/s08/stm8/pdk14/pdk15 3.9.0 #11195 (Linux)
published under GNU General Public License (GPL)
[frans-willem@Hydra controller]$ cat main.h | grep ROTOR_ANGLE
#define MOTOR_ROTOR_ANGLE_90                      73
#define MOTOR_ROTOR_ANGLE_150                     116
#define MOTOR_ROTOR_ANGLE_210                     158
#define MOTOR_ROTOR_ANGLE_270                     201
#define MOTOR_ROTOR_ANGLE_330                     243
#define MOTOR_ROTOR_ANGLE_30                      30
[frans-willem@Hydra controller]$ rm -rf main.hex
[frans-willem@Hydra controller]$ make -f Makefile_linux >/dev/null 2>&1
[frans-willem@Hydra controller]$ cp main.hex main_constants.hex
[frans-willem@Hydra controller]$ git checkout main.h
Updated 1 path from the index
[frans-willem@Hydra controller]$ cat main.h | grep ROTOR_ANGLE
#define MOTOR_ROTOR_ANGLE_90                      (63  + MOTOR_ROTOR_OFFSET_ANGLE)
#define MOTOR_ROTOR_ANGLE_150                     (106 + MOTOR_ROTOR_OFFSET_ANGLE)
#define MOTOR_ROTOR_ANGLE_210                     (148 + MOTOR_ROTOR_OFFSET_ANGLE)
#define MOTOR_ROTOR_ANGLE_270                     (191 + MOTOR_ROTOR_OFFSET_ANGLE)
#define MOTOR_ROTOR_ANGLE_330                     (233 + MOTOR_ROTOR_OFFSET_ANGLE)
#define MOTOR_ROTOR_ANGLE_30                      (20  + MOTOR_ROTOR_OFFSET_ANGLE)
[frans-willem@Hydra controller]$ rm -rf main.hex
[frans-willem@Hydra controller]$ make -f Makefile_linux >/dev/null 2>&1
[frans-willem@Hydra controller]$ cp main.hex main_expressions.hex
[frans-willem@Hydra controller]$ diff -s main_constants.hex main_expressions.hex
Files main_constants.hex and main_expressions.hex are identical
```

Comment (leon927) [reply src/controller/main.h]: That is great actually and thank you for taking the time to test it out! Will fix in the future!

File: src/display/KT-LCD3/lcd.c [hunk 10]
```diff
@@ -3413,226 +3413,142 @@ void lcd_print(uint32_t ui32_number, uint8_t ui8_lcd_field, uint8_t ui8_options)
 
 void lcd_enable_w_symbol (uint8_t ui8_state)
 {
-  if (ui8_state)
-    ui8_lcd_frame_buffer[9] |= 128;
-  else
-    ui8_lcd_frame_buffer[9] &= ~128;
+  if (ui8_state) { ui8_lcd_frame_buffer[9] |= 128; }
```
Comment (Frans-Willem) [inline src/display/KT-LCD3/lcd.c:3583]: Have you actually tested this change ?
I'm about 95% certain you don't intend to remove the else clause ;)
If you are dead-set on a one-liner, try this:
```
ui8_lcd_frame_buffer[9] = (ui8_lcd_frame_buffer[9] & ~128) | (ui8_state << 7)
```
(I'd actually be curious what ends up smaller in code size or faster, a compare + or / and, or a and + shift + or...)

Comment (leon927) [reply src/display/KT-LCD3/lcd.c:3583]: Yes of course and the way the firmware controls the LCD it is allowed :) It is such a small difference but every little bit matters.

No idea what ends up smaller or faster but I love those kind of questions! Do you have a test planned?
